### ISSUE TITLE
SignatureHelp doesn't respect documentationFormat

### ISSUE THREAD

**Issue by hassec** [clangd/clangd#945]
Whether `documentationFormat` is set to `["markdown", "plaintext"]`, `["markdown"]`, or just `["plaintext']` the documentation is always returned as plain string inside the `SignatureHelp` response.       
This behavior is different from the documentation returned inside a completion request.

**Example Code** 

```cpp
#include<functional>
int main(){
  int a{1};
  int b{2};

  std::swap(a, b);
//             |---- cursor is here on the b when triggering signatureHelp
}
```

**Logs**
Startup message with full capabilities, contains `"documentationFormat" : ["markdown"]`
```
[ERROR][2021-12-07 22:40:34] .../vim/lsp/rpc.lua:417	"rpc"	"clangd"	"stderr"	"I[22:40:34.233] clangd version 13.0.0\nI[22:40:34.233] Features: linux\nI[22:40:34.233] PID: 38703\nI[22:40:34.233] Working directory: /home/christoph\nI[22:40:34.233] argv[0]: clangd\nI[22:40:34.233] argv[1]: --background-index\nI[22:40:34.233] argv[2]: -completion-style=detailed\nI[22:40:34.233] argv[3]: -header-insertion=never\nI[22:40:34.233] argv[4]: --log=verbose\nV[22:40:34.233] User config file is /home/christoph/.config/clangd/config.yaml\nI[22:40:34.233] Starting LSP over stdin/stdout\nV[22:40:34.233] <<< {\"id\":1,\"jsonrpc\":\"2.0\",\"method\":\"initialize\",\"params\":{\"capabilities\":{\"callHierarchy\":{\"dynamicRegistration\":false},\"offsetEncoding\":[\"utf-8\",\"utf-16\"],\"textDocument\":{\"codeAction\":{\"codeActionLiteralSupport\":{\"codeActionKind\":{\"valueSet\":[\"\",\"Empty\",\"QuickFix\",\"Refactor\",\"RefactorExtract\",\"RefactorInline\",\"RefactorRewrite\",\"Source\",\"SourceOrganizeImports\",\"quickfix\",\"refactor\",\"refactor.extract\",\"refactor.inline\",\"refactor.rewrite\",\"source\",\"source.organizeImports\"]}},\"dataSupport\":true,\"dynamicRegistration\":false,\"resolveSupport\":{\"properties\":[\"edit\"]}},\"completion\":{\"completionItem\":{\"commitCharactersSupport\":true,\"deprecatedSupport\":true,\"documentationFormat\":[\"markdown\",\"plaintext\"],\"insertReplaceSupport\":true,\"labelDetailsSupport\":true,\"preselectSupport\":true,\"resolveSupport\":{\"properties\":[\"documentation\",\"detail\",\"additionalTextEdits\"]},\"snippetSupport\":true,\"tagSupport\":{\"valueSet\":[1]}},\"completionItemKind\":{\"valueSet\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]},\"contextSupport\":false,\"dynamicRegistration\":false,\"editsNearCursor\":true},\"declaration\":{\"linkSupport\":true},\"definition\":{\"linkSupport\":true},\"documentHighlight\":{\"dynamicRegistration\":false},\"documentSymbol\":{\"dynamicRegistration\":false,\"hierarchicalDocumentSymbolSupport\":true,\"symbolKind\":{\"valueSet\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},\"hover\":{\"contentFormat\":[\"markdown\",\"plaintext\"],\"dynamicRegistration\":false},\"implementation\":{\"linkSupport\":true},\"publishDiagnostics\":{\"relatedInformation\":true,\"tagSupport\":{\"valueSet\":[1,2]}},\"references\":{\"dynamicRegistration\":false},\"rename\":{\"dynamicRegistration\":false,\"prepareSupport\":true},\"signatureHelp\":{\"dynamicRegistration\":false,\"signatureInformation\":{\"activeParameterSupport\":true,\"documentationFormat\":[\"markdown\"],\"parameterInformation\":{\"labelOffsetSupport\":true}}},\"synchronization\":{\"didSave\":true,\"dynamicRegistration\":false,\"willSave\":false,\"willSaveWaitUntil\":false},\"typeDefinition\":{\"linkSupport\":true}},\"window\":{\"showDocument\":{\"support\":false},\"showMessage\":{\"messageActionItem\":{\"additionalPropertiesSupport\":false}},\"workDoneProgress\":true},\"workspace\":{\"applyEdit\":true,\"configuration\":true,\"symbol\":{\"dynamicRegistration\":false,\"hierarchicalWorkspaceSymbolSupport\":true,\"symbolKind\":{\"valueSet\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},\"workspaceEdit\":{\"resourceOperations\":[\"rename\",\"create\",\"delete\"]},\"workspaceFolders\":true}},\"clientInfo\":{\"name\":\"Neovim\",\"version\":\"0.6.0\"},\"initializationOptions\":{},\"processId\":38700,\"rootPath\":null,\"rootUri\":null,\"trace\":\"off\",\"workspaceFolders\":null}}\n\nI[22:40:34.233] <-- initialize(1)\nI[22:40:34.233] --> reply:initialize(1) 0 ms\nV[22:40:34.233] >>> {\"id\":1,\"jsonrpc\":\"2.0\",\"result\":{\"capabilities\":{\"astProvider\":true,\"callHierarchyProvider\":true,\"codeActionProvider\":{\"codeActionKinds\":[\"quickfix\",\"refactor\",\"info\"]},\"compilationDatabase\":{\"automaticReload\":true},\"completionProvider\":{\"allCommitCharacters\":[\" \",\"\\t\",\"(\",\")\",\"[\",\"]\",\"{\",\"}\",\"<\",\">\",\":\",\";\",\",\",\"+\",\"-\",\"/\",\"*\",\"%\",\"^\",\"&\",\"#\",\"?\",\".\",\"=\",\"\\\"\",\"'\",\"|\"],\"resolveProvider\":false,\"triggerCharacters\":[\".\",\"<\",\">\",\":\",\"\\\"\",\"/\"]},\"declarationProvider\":true,\"definitionProvider\":true,\"documentFormattingProvider\":true,\"documentHighlightProvider\":true,\"documentLinkProvider\":{\"resolveProvider\":false},\"documentOnTypeFormattingProvider\":{\"firstTriggerCharacter\":\"\\n\",\"moreTriggerCharacter\":[]},\"documentRangeFormattingProvider\":true,\"documentSymbolProvider\":true,\"executeCommandProvider\":{\"commands\":[\"clangd.applyFix\",\"clangd.applyTweak\"]},\"hoverProvider\":true,\"implementationProvider\":true,\"memoryUsageProvider\":true,\"referencesProvider\":true,\"renameProvider\":{\"prepareProvider\":true},\"selectionRangeProvider\":true,\"semanticTokensProvider\":{\"full\":{\"delta\":true},\"legend\":{\"tokenModifiers\":[\"declaration\",\"deprecated\",\"deduced\",\"readonly\",\"static\",\"abstract\",\"dependentName\",\"defaultLibrary\",\"functionScope\",\"classScope\",\"fileScope\",\"globalScope\"],\"tokenTypes\":[\"variable\",\"variable\",\"parameter\",\"function\",\"method\",\"function\",\"property\",\"variable\",\"class\",\"interface\",\"enum\",\"enumMember\",\"type\",\"type\",\"unknown\",\"namespace\",\"typeParameter\",\"concept\",\"type\",\"macro\",\"comment\"]},\"range\":false},\"signatureHelpProvider\":{\"triggerCharacters\":[\"(\",\",\"]},\"textDocumentSync\":{\"change\":2,\"openClose\":true,\"save\":true},\"typeHierarchyProvider\":true,\"workspaceSymbolProvider\":true},\"offsetEncoding\":\"utf-8\",\"serverInfo\":{\"name\":\"clangd\",\"version\":\"clangd version 13.0.0 linux x86_64-pc-linux-gnu\"}}}\n\n"
```
**Reply**
```
{
  activeParameter = 1,
  activeSignature = 0,
  signatures = { {
      documentation = "\n  @brief Swaps two values.\n  @param  __a  A thing of arbitrary type.\n  @param  __b  Another thing of arbitrary type.\n  @return   Nothing.\n  ",
      label = "swap(_Tp &__a, _Tp &__b) -> typename enable_if<__and_<__not_<__is_tuple_like<_Tp>>, is_move_constructible<_Tp>, is_move_assignable<_Tp>>::value>::type",
      parameters = { {
          label = { 5, 13 }
        }, {
          label = { 15, 23 }
        } }
    } }
}
```

**System information**

Output of `clangd --version`:
```
clangd version 13.0.0
Features: linux
Platform: x86_64-pc-linux-gnu
```

Editor/LSP plugin: neovim + [nvim-cmp](https://github.com/hrsh7th/nvim-cmp) + [cmp-nvim-lsp](https://github.com/hrsh7th/cmp-nvim-lsp)

Operating system: Arch Linux

**Comment by kirillbobyrev** [comment]
> This behavior is different from the documentation returned inside a completion request.

Can you please show the different documentation format in the completion request? My understanding is, it should be the same (I also tried `std::swap` in my editors: only the header is escaped in ``, the reset seems to be the same plain text.

In general, this seems to be an issue with our Doxygen parsing (-> duplicate of https://github.com/clangd/clangd/issues/529), please let me know if this isn't the case.

**Comment by hassec** [comment]
Thanks for the quick reply :blush: 

of course, here is the completion item of the corresponding overload:
(*Note*: this is now from running with `--completion-style=detailed` to get individual overloads) 

`documentationFormat` is in neovim per default set to `["markdown", "plaintext"]` so `clangd` is doing the correct thing and returns markdown with everything properly escaped:

```
{
  detail = "typename enable_if<__and_<__not_<__is_tuple_like<_Tp>>, is_move_constructible<_Tp>, is_move_assignable<_Tp>>::value>::type",
  documentation = {
    kind = "markdown",
    value = "From `<type_traits>`  \n@brief Swaps two values.  \n@param \\_\\_a A thing of arbitrary type.  \n@param \\_\\_b Another thing of arbitrary type.  \n@return Nothing."
  },
  filterText = "swap",
  insertText = "swap",
  insertTextFormat = 1,
  kind = 3,
  label = " swap(_Tp &__a, _Tp &__b)",
  score = 1.5728549957275,
  sortText = "3fb6acb0swap",
  textEdit = {
    newText = "swap",
    range = {
      end = {
        character = 11,
        line = 7
      },
      start = {
        character = 7,
        line = 7
      }
    }
  }
}
```

If I set the `documentationFormat` to `plaintext` for the completion, I get (no escapes):
```
documentation = {
  kind = "plaintext",
  value = "From <type_traits>\n@brief Swaps two values.\n@param __a A thing of arbitrary type.\n@param __b Another thing of arbitrary type.\n@return Nothing."
}
```

So the problem here isn't as much that the Doxygen keywords don't get parsed, but the fact that I can set `documentationFormat = 'markdown'` for the signature help but still receive a string that isn't properly escaped.

**Comment by kirillbobyrev** [comment]
Ah, I see, makes sense, thank you!

**Comment by kadircet** [comment]
So just to be clear, what clangd does here is in accordance with the [LSP](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#signatureInformation).

The `documentation` field of a `SignatureInformation` can be either a `string` or a `MarkupContent`, clangd provides a `string` for `signaturehelp` as we don't have any fancy markup to offer. The client should not be treating a string typed value in a `SignatureInformation.documentation` field, no matter what. I believe you should file an issue against your LSP client plugin.

We can in future consider updating clangd to provide a `MarkupContent` object when we have some extra information to provide in the documentation of signature help, but today we don't.

**Comment by hassec** [comment]
@kadircet so this might be a shortcoming of my understanding of the LSP. 

I was aware that `string` and `MarkupContent` are allowed return values, but figured setting `documentationFormat` to `markdown` and not getting any warning or whatever means the LSP will reply with `markdown`

So does a reply of type `string` simply mean that the client should assume that any setting of `documentationFormat` is ignored, and thus does `documentationFormat` only ever apply to a return of type `MarkupContent` ? 

>  I believe you should file an issue against your LSP client plugin.    
 
I think this is disconnected of any client, but more about whether `documentationFormat` should apply to a `string` return type. I guess that isn't the case. 

> We can in future consider updating clangd to provide a MarkupContent object when we have some extra information to provide in the documentation of signature help, but today we don't.

Well, it's the identical documentation that is provided in the `completionItem` so why wouldn't it make sense to also turn it into a `MarkupContent` here?

**Comment by kadircet** [comment]
> I was aware that string and MarkupContent are allowed return values, but figured setting documentationFormat to markdown and not getting any warning or whatever means the LSP will reply with markdown

The information sent with capabilities just indicates that the client can deal with certain types of data and prefers to see them in some order. It isn't very precise about what to do in a case which client and server don't have a common data type.
`string` typed value fills in that gap to some extent. it's just data without any meaning, without additional information (like a secret agreement between the server and the client) client cannot assume it has markdown or any other semantics. Because there's a dedicated type (`MarkupContent`) for indicating those semantics.

> So does a reply of type string simply mean that the client should assume that any setting of documentationFormat is ignored, and thus does documentationFormat only ever apply to a return of type MarkupContent ?

As described above, the client should just interpret a `documentation` of type `string` as raw-data, and not look for markdown or any other semantics.

> I think this is disconnected of any client

Sorry I assumed you were having a suboptimal experience. Hence suggested filing an issue there.

> whether documentationFormat should apply to a string return type. I guess that isn't the case.

Yes, as described, there's a specific `MarkupContent` type for indicating that semantic.

> Well, it's the identical documentation that is provided in the completionItem so why wouldn't it make sense to also turn it into a MarkupContent here?

Well thanks for bringing it up again, apparently I was too fast to react on this one.

In theory the contents in two cases are not identical. When we know about the particular header a symbol is coming from, we chose to include that in code completion documentation (to inform the user about a possible insertion being triggered) and we do that in markdown format:
```
From `<header.h>`
```
We don't do it for signature help, because we are not completing anything so we can't trigger insertion and user already has the symbol so this is probably just going to be noise.

I thought that was the **only** markdown item we've introduced hence no need for making signature help more complicated.

But apparently it isn't the case. During code completion we also parse the comment for backticks (``` ` ```) and introduce text between as inline code blocks rather than plain text. There's no reason for not doing this in signaturehelp documentations too. I'll send out a patch.
